### PR TITLE
Rule8525

### LINKED ISSUE #18: Rule 8525 : Either Date of Birth or Expected Date of Birth must be provided (but not both)
Either

a) \<PersonBirthDate\> (N00066) must be present  and \<ExpectedPersonBirthDate\> (N00098) must be omitted; or

b) \<ExpectedPersonBirthDate\> (N00098) must be present and \<PersonBirthDate\> (N00066) must be omitted.

### PR DESCRIPTION
**Pull request by willLPD2I**
Tinas rule
closes #18 

Tambe, I want to talk about how to do the link ID for this one. The rule checks if either PersonBirthDate or ExpectedPersonBirthDate are NaNs, and returns rows where either both are NaN or neither are NaN. This means that where both are NaN, if we are using affected fields, the link ID is LAchildID, NaN, NaN, which is not good!

### CODE REVIEW

**Comment by tab1tha** [comment]
@willLPD2I You are right. (LAchildID, NaN, NaN) is not good. 
In the end, LAchildID will be the only distinguishing factor between the (LAchildID, NaN, NaN) produced by one row and that produced by a totally unrelated row.
Using LAchildID as the ERROR_ID seems to be our best bet in this case. 

That makes me think of the conversation I had with Stephen last week about his rule. I'll give that a look again.

**Comment by willLPD2I** [comment]
@tab1tha I changed the link_id, this should be good to merge if you can just pop a review in

File: cin_validator/rules/cin2022_23/rule_8525.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            {
+                "LAchildID": "child5",
+                "PersonBirthDate": "26/05/2000",
+                "ExpectedPersonBirthDate": "25/05/2000",
+            },  # Fails because both DOB and expected DOB are present
+            {
+                "LAchildID": "child6",
+                "PersonBirthDate": pd.NA,
+                "ExpectedPersonBirthDate": pd.NA,
+            },  # Fails becasue there is no DOB or expected DOB
```
Comment (tab1tha) [inline cin_validator/rules/cin2022_23/rule_8525.py:112]: little typo

File: cin_validator/rules/cin2022_23/rule_8525.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    condition_1 = (df[PersonBirthDate].isna() == True) & (
+        df[ExpectedPersonBirthDate].isna() == True
+    )
+    condition_2 = df[PersonBirthDate].notna() & df[ExpectedPersonBirthDate].notna()
+
+    # get all the data that fits the failing condition. Reset the index so that ROW_ID now becomes a column of df
+    df_issues_1 = df[condition_1]
+    df_issues_2 = df[condition_2]
+    df_issues = pd.concat([df_issues_1, df_issues_2]).reset_index()
+
```
Comment (tab1tha) [inline cin_validator/rules/cin2022_23/rule_8525.py:53]: lines 50, 51 and 52 can be summarised into

df_issues = df[condition_1 | condition_2].reset_index()

**Review by tab1tha — CHANGES_REQUESTED** [review]

File: cin_validator/rules/cin2022_23/rule_8525.py (hunk 1 above)
Comment (willLPD2I) [reply cin_validator/rules/cin2022_23/rule_8525.py:112]: i fully can't work out what the typo is

File: cin_validator/rules/cin2022_23/rule_8525.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    # Generate a unique ID for each instance of an error. In this case,
+    # - If only LAchildID is used as an identifier, multiple instances of the error on a child will be understood as 1 instance.
+    # We don't want that because in reality, a child can have multiple instances of an error.
+    # - If we use the LAchildID-CPPstartDate combination, that artificially cancels out the instances where a start date repeats for the same child.
+    # Another rule checks for that condition. Not this one.
+    # - It is very unlikely that a combination of LAchildID-CPPstartDate-CPPendDate will repeat in the DataFrame.
+    # Hence, it can be used as a unique identifier of the row.
+
+    # Replace CPPstartDate and CPPendDate below with the columns concerned in your rule.
+    link_id = tuple(
```
[the remark this answers is not in the document]
Comment (tab1tha) [reply cin_validator/rules/cin2022_23/rule_8525.py:64]: @willLPD2I

**Review by tab1tha — CHANGES_REQUESTED** [review]

File: cin_validator/rules/cin2022_23/rule_8525.py (hunk 2 above)
Comment (willLPD2I) [reply cin_validator/rules/cin2022_23/rule_8525.py:51]: done

File: cin_validator/rules/cin2022_23/rule_8525.py (hunk 1 above)
Comment (willLPD2I) [reply cin_validator/rules/cin2022_23/rule_8525.py]: done

File: cin_validator/rules/cin2022_23/rule_8525.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    df.index.name = "ROW_ID"
+
+    # lOGIC
+    # Implement rule logic as described by the Github issue.
+    # Put the description as a comment above the implementation as shown.
+
+    # Either Date of Birth or Expected Date of Birth must be provided (but not both)
+    # condition_1 = (df[PersonBirthDate].isna() & df[ExpectedPersonBirthDate].isna())
+    condition_1 = (df[PersonBirthDate].isna() == True) & (
+        df[ExpectedPersonBirthDate].isna() == True
```
[the remark this answers is not in the document]
Comment (willLPD2I) [reply cin_validator/rules/cin2022_23/rule_8525.py:45]: I know there was a reason I did it like this with Tina, I think it was as a teaching exercise in the workshop and we forgot to remove it. I've made the change now.

**Review by tab1tha — APPROVED** [review]

**Pull request merged**